elasticdl/python/allreduce/tensorflow_controller.py

```diff
@@ -74,3 +75,44 @@ def train_one_batch_with_retries(self, func, *args, **kwargs):
                     self._rendezvous_manager.init_horovod_if_needed()
         self.data_shard_service.report_batch_done()
         return result
+
+
+class TensorFlowV1AllReduceController(AllReduceController):
+    """The controller is responsible for elastic training of
+    TensorFlow eager execution using AllReduce.
+    """
+
+    def __init__(self, master_client, master_addr):
+        super(TensorFlowV1AllReduceController, self).__init__(
+            master_client, master_addr
+        )
+        self._bcast_op = None
+
+    def broadcast(self):
+        if self._bcast_op is None:
+            self._variables = tf.global_variables()
+            self._bcast_op = broadcast_variables(self._variables, root_rank=0)
+        session = tf.get_default_session()
+        session.run(self._bcast_op)
+
+    def train_one_batch_with_retries(self, func, *args, **kwargs):
+        for _ in range(DEFAULT_MAX_ALLREDUCE_RETRY_NUM + 1):
+            try:
+                self._broadcast_if_needed()
+                result = func(*args, **kwargs)
+                return result
+            except UnknownError as e:
+                logger.warning(
+                    "Failed to perform allreduce operation on "
+                    "the gradients. Retrying..."
+                )
+                # Those error message show that the communication
+                # to merge gradient fails and we can rebuild the
+                # communication.
+                if (
```

Comment:
How about the `HorovodUpdateError` and `HorovodInternalError`?

Reply:
`HorovodUpdateError` and `HorovodInternalError` are the errors in Horovod not TensorFlow. Horovod will convert the `UnknownError` in TensorFlow to those errors.
https://github.com/horovod/horovod/blob/e94d8ea0dff8c2b45698cfe4fabb2e6553d0b9a8/horovod/tensorflow/elastic.py#L53-L61

Here, we get the error from TensorFlow, not Horovod. So we can not use `HorovodUpdateError` and `HorovodInternalError`.